keps/sig-storage/177-volume-snapshot-validation/tighten-validation-webhook-crd.md

```diff
@@ -0,0 +1,799 @@
+
+### Authentication
+
+There are two directions to authentication. Authenticating the webhook server is legitimate, and authenticating the k8s api server is legitimate. 
+
+The API server authenticates the webhook server through TLS certificates and HTTPS. This is required, and an example method of deploying the webhook server with HTTPS will be provided.
+
+Authentication on incoming requests to the webhook server is configurable however out of scope of this document. It’s the user’s responsibility in general to configure the webhook service and the API server if authentication is required(details). The web server implementation, however, should allow users to configure whether authentication is required or not. If no authentication config is specified, the webhook server should default to “NoClientCert”, which effectively will not authenticate the identity of the clients.
+
+### Timeout
+
+Webhooks add latency to each API server call, thus setting up a reasonable timeout for each AdmissionReview request from the webhook server side is critical. The default timeout is 10 seconds if not specified. When an AdmissionReview request sent to the webhook server timed out, `failurePolicy`(default to `Fail` which is equivalent to disallow) will be triggered.
+
+In the ValidatingWebhookConfiguration yaml example, a default timeout of two seconds is provided, cluster admins who wish to change the timeout may change the value of `timeoutSeconds`
+
+To avoid migration pain it is recommended to start with a `failurePolicy` value of `Ignore`, changing it to `Fail` only after the webhook is confirmed to have been installed successfully.
+
+### Idempotency/Deadlock
+
+Since only validating webhooks will be introduced in this version, idempotency/deadlock are not relevant.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+CA can deploy the webhook server.
+Users can create and update snapshot objects with confidence invalid updates will be rejected.
+
+Following are some typical scenarios we are aiming to prevent:
+- Creation of invalid CRs
+- Reject if a VolumeSnapshot CR does not have a legit VolumeSnapshotSource, i.e., missing both PersistentVolumeName and VolumeSnapshotContentName.
+- Reject if a VolumeSnapshotContent CR does not have a legit VolumeSnapshotContentSource, i.e., both VolumeHandle and SnapshotHandle have been specified
+- Updating immutable fields
+- Reject updates to VolumeSnapshot’s VolumeSnapshotSource
+- Reject updates to VolumeSnapshotContent’s VolumeSnapshotContentSource
+- Reject updates to VolumeSnapshotContent’s volume snapshot ref after binding
```

Comment:
This KEP talks a lot about webhooks and tightening validation, but it does not actually list *what* validations are planed to be added. Is this complete list? What fields should be immutable?

Reply:
Please take a look at https://github.com/kubernetes/enhancements/blob/ae1cfab09d59aafff22f7147c0a5e6b5c6cb38b8/keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md#proposal

The list is there and also repeated in the table at https://github.com/kubernetes/enhancements/blob/ae1cfab09d59aafff22f7147c0a5e6b5c6cb38b8/keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md#validating-scenarios